components/script/dom/node.rs

```diff
@@ -261,14 +280,33 @@ impl Node {
         }
 
         new_child.parent_node.set(Some(self));
+        if let Some(shadow_root) = self.downcast::<ShadowRoot>() {
+            new_child
+                .composed_parent_node
+                .set(Some(shadow_root.Host().upcast::<Node>()));
+        } else {
+            new_child.composed_parent_node.set(Some(self));
+        }
         self.children_count.set(self.children_count.get() + 1);
 
         let parent_in_doc = self.is_in_doc();
-        for node in new_child.traverse_preorder() {
+        let parent_in_shadow_tree = self.is_in_shadow_tree();
+        let parent_is_connected = self.is_connected();
+
+        for node in new_child.traverse_preorder(ShadowIncluding::No) {
+            if parent_in_shadow_tree {
+                if let Some(shadow_root) = self.downcast::<ShadowRoot>() {
+                    node.set_owner_shadow_root(&*shadow_root);
+                } else if let Some(shadow_root) = self.owner_shadow_root() {
+                    node.set_owner_shadow_root(&*shadow_root);
+                }
```

Comment:
We should assert that `owner_shadow_root` is `Some` here, rather than doing nothing if there's broken state.